### PR TITLE
feature/oidcs-based-authentication

### LINKED ISSUE #59: authentication

### PR DESCRIPTION
**Pull request by RabbITCybErSeC**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,3 +1,9 @@
 SOARCA_URI: http://localhost:8080
+SOARCA_GUI_URI: http://localhost:8081
```
Comment (MaartendeKruijf) [inline .env.example]: Maybe include port not in this URL as you have the port variable already

File: auth/auth.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	"soarca-gui/utils"
+	"strconv"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/gorilla/securecookie"
+	"golang.org/x/oauth2"
+)
+
+const (
+	OIDC_CALL_BACK_PATH          = "/oidc-callback"
```
Comment (MaartendeKruijf) [inline auth/auth.go]: Maybe make this optional variables as well, this to allow the user to set a path.

File: auth/auth.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	COOKIE_ENCRYPTION_KEY_LENGTH = 32
+	COOKIE_SECRET_KEY_LENGHT     = 32
```
Comment (MaartendeKruijf) [inline auth/auth.go]: Maybe lat the user specify longer keys if they wish?

File: auth/auth.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+		client := &http.Client{Transport: tr}
+		log.Println("Warning: TLS verification is disabled. This should not be used in production.")
+		return client
+	}
+	return http.DefaultClient
+}
+
+func NewAuthenticator(cj cookies.ICookieJar, OIDCconfig *oidc.Config, OauthConfig *oauth2.Config, verifierProvider *oidc.Provider, skipTLSVerify bool, userClaims *UserClaimsConfig) *Authenticator {
```
Comment (MaartendeKruijf) [inline auth/auth.go]: New()

File: auth/auth.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+type Authenticator struct {
+	Cookiejar        cookies.ICookieJar
+	OIDCconfig       *oidc.Config
+	OauthConfig      *oauth2.Config
+	verifierProvider *oidc.Provider
+	userclaimConfig  *UserClaimsConfig
+	skipTLSVerify    bool
+}
+
+func SetupOIDCAuthHandler() *Authenticator {
```
Comment (MaartendeKruijf) [inline auth/auth.go]: Why is this function here and the NewAuthenticator?

File: auth/context.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	}
+	return user, true
+}
+
+func GetUserPermissions(gc *gin.Context) []string {
+	permissions, exists := gc.Get(permissionsContextKey)
+	if !exists {
+		return []string{}
+	}
+	groups, ok := permissions.([]string)
```
Comment (MaartendeKruijf) [inline auth/context.go]: Please don't cast

File: auth/context.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	}
+	var user User
+	err := json.Unmarshal([]byte(userString), &user)
+	if err != nil {
+		return User{}, false
+	}
+	return user, true
+}
+
+func GetUserPermissions(gc *gin.Context) []string {
```
Comment (MaartendeKruijf) [inline auth/context.go]: Permissions is not what is returned here. Either call the function GetUserAssignedGroups or actually return the permissions.

File: auth/auth.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+	}
+	if env.clientID == "" {
+		log.Fatal("invalid oidc client ID for the env: OIDC_CLIENT_ID")
+	}
+	if env.clientSecret == "" {
+		log.Fatal("invalid oidc client secret for the env: OIDC_CLIENT_SECRET")
+	}
+}
+
+func setupHTTPClient(skipTLS bool) *http.Client {
```
Comment (MaartendeKruijf) [inline auth/auth.go]: skipTLS is not correct is should be skipTlsValidation

**Review by MaartendeKruijf — CHANGES_REQUESTED** [review]

File: auth/auth.go (hunk 3 above)
Comment (RabbITCybErSeC) [reply auth/auth.go]: This is already the strongest key: as specified in the gorilla session  docs:

It is recommended to use an authentication key with 32 or 64 bytes. The encryption key, if set, must be either 16, 24, or 32 bytes to select AES-128, AES-192, or AES-256 modes. 

https://pkg.go.dev/github.com/gorilla/sessions#NewCookieStore

File: go.mod [hunk 9]
```diff
@@ -1,16 +1,24 @@
 module soarca-gui
 
-go 1.22.2
+go 1.23
+
+toolchain go1.23.1
```
Comment (MaartendeKruijf) [inline go.mod:5]: Why is this added

File: handlers/status.go [hunk 10]
```diff
@@ -1,25 +1,38 @@
...
-func NewStatusHandler(backend backend.Status) statusHandler {
-	return statusHandler{status: backend}
+func (s *statusHandler) fetchStatus(context *gin.Context) (string, error) {
+	if s.authenticated {
+		bearerToken, exists := gauth_context.GetTokenFromContext(context)
+		if exists {
+			return s.status.GetPongFromStatus(bearerToken)
+		}
+	}
+	return s.status.GetPongFromStatus("")
```
Comment (MaartendeKruijf) [inline handlers/status.go]:
```suggestion
	bearerToken, _ := gauth_context.GetTokenFromContext(context)
	return s.status.GetPongFromStatus(bearerToken)
```

File: routes/routes.go [hunk 11]
```diff
@@ -18,34 +21,65 @@ func Setup(app *gin.Engine) {
 		ctx.Redirect(http.StatusTemporaryRedirect, "/404-page")
 	})
 
-	reporter := soarca.NewReport(utils.GetEnv("SOARCA_URI", "http://localhost:8080"), &http.Client{})
+	authEnabled, _ := strconv.ParseBool(utils.GetEnv("AUTH_ENABLED", "false"))
+	authEnabledToSoarca, _ := strconv.ParseBool(utils.GetEnv("AUTH_ENABLED", "false"))
```
Comment (MaartendeKruijf) [inline routes/routes.go]: Why split this?

File: go.mod [hunk 12]
```diff
@@ -1,16 +1,24 @@
...
 
 require (
+	github.com/COSSAS/gauth v1.0.1
```
Comment (MaartendeKruijf) [inline go.mod]:
```suggestion
	github.com/COSSAS/gauth v1.0.2
```

**Review by MaartendeKruijf — CHANGES_REQUESTED** [review]

File: handlers/oidc_handler.go [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+	authenticator *gauth.Authenticator
+}
+
+func NewOIDCAuthHandler(authenticator *gauth.Authenticator) *OIDCAuthHandler {
+	return &OIDCAuthHandler{authenticator: authenticator}
+}
+
+func (auth *OIDCAuthHandler) OIDCAuthPageHandler(context *gin.Context) {
+	// context.Header("HX-Redirect", "/dashboard")
+	// context.String(http.StatusFound, "")
```
Comment (MaartendeKruijf) [inline handlers/oidc_handler.go]: remove unused code

File: handlers/reporting.go [hunk 14]
```diff
@@ -3,91 +3,115 @@ package handlers
...
+	return reportingHandler{
+		reporter:      backend,
+		authenticated: authenticated,
+	}
 }
 
-func (r *reportingHandler) ReportingIndexHandler(context *gin.Context) {
-	render := utils.NewTempl(context, http.StatusOK, reporting.ReportingIndex())
-	context.Render(http.StatusOK, render)
+func (r *reportingHandler) fetchReports(context *gin.Context) ([]reporter.PlaybookExecutionReport, error) {
```
Comment (MaartendeKruijf) [inline handlers/reporting.go]:
```suggestion
func (reporting *reportingHandler) fetchReports(context *gin.Context) ([]reporter.PlaybookExecutionReport, error) {
```

File: handlers/status.go [hunk 15]
```diff
@@ -1,25 +1,36 @@
...
-	return statusHandler{status: backend}
+func (s *statusHandler) fetchStatus(context *gin.Context) (string, error) {
+	if s.authenticated {
+		bearerToken, _ := gauth_context.GetTokenFromContext(context)
+		return s.status.GetPongFromStatus(bearerToken)
+	}
+	return s.status.GetPongFromStatus("")
 }
 
 func (s *statusHandler) HealthComponentHandler(context *gin.Context) {
```
Comment (MaartendeKruijf) [inline handlers/status.go]:
```suggestion
func (status *statusHandler) HealthComponentHandler(context *gin.Context) {
```

File: routes/routes.go [hunk 16]
```diff
@@ -18,34 +21,64 @@ func Setup(app *gin.Engine) {
 		ctx.Redirect(http.StatusTemporaryRedirect, "/404-page")
 	})
 
-	reporter := soarca.NewReport(utils.GetEnv("SOARCA_URI", "http://localhost:8080"), &http.Client{})
+	authEnabled, _ := strconv.ParseBool(utils.GetEnv("AUTH_ENABLED", "false"))
 
-	status := soarca.NewStatus(utils.GetEnv("SOARCA_URI", "http://localhost:8080"), &http.Client{})
+	reporter := soarca.NewReport(utils.GetEnv("SOARCA_URI", "http://localhost:8080"), &http.Client{}, authEnabled)
+	status := soarca.NewStatus(utils.GetEnv("SOARCA_URI", "http://localhost:8080"), &http.Client{}, authEnabled)
```
Comment (MaartendeKruijf) [inline routes/routes.go:27]: Maybe wrap endpoint config into a struct so it can be passed easily.

File: routes/routes.go [hunk 17]
```diff
@@ -18,34 +21,64 @@ func Setup(app *gin.Engine) {
...
 
+	auth, err := gauth.New(gauth.OIDCRedirectConfig())
+	authHandler := handlers.NewOIDCAuthHandler(auth)
+	if err != nil {
+		log.Fatal("could not configure oidc redirect config: ", err)
+	}
 	publicRoutes := app.Group("/")
+	if err != nil {
+		log.Fatal("AUTH_ENABLED flag could not be parsed properly should be 'true' | 'false'")
+	}
```
Comment (MaartendeKruijf) [inline routes/routes.go]: unreachable if err != nil

File: routes/routes.go [hunk 18]
```diff
@@ -18,34 +21,64 @@ func Setup(app *gin.Engine) {
...
+	protectedRoutes := app.Group("/")
+	protectedRoutes.Use(auth.LoadAuthContext())
+	if authEnabled {
+		PublicOIDCRoutes(publicRoutes, authHandler)
+	} else {
+		PublicRoutes(publicRoutes)
+	}
```
Comment (MaartendeKruijf) [inline routes/routes.go]: Shouldn't this be handle in the called function

File: routes/routes.go [hunk 19]
```diff
@@ -18,34 +21,64 @@ func Setup(app *gin.Engine) {
...
 	publicRoute := app.Group("/")
 	{
 		publicRoute.GET("/", authHandler.AuthPage)
 		publicRoute.POST("/login", authHandler.Login)
-		publicRoute.GET("/dashboard", handlers.HomeDashboard)
 
 	}
-
 	publicRoute.StaticFS("/public", public.GetPublicAssetsFileSystem())
 }
```
Comment (MaartendeKruijf) [inline routes/routes.go]: Duplicate code

**Review by MaartendeKruijf — CHANGES_REQUESTED** [review]

**Review by MaartendeKruijf — APPROVED** [review]

**Pull request merged**